zsh-history-substring-search.zsh

```diff
@@ -233,20 +237,33 @@ _history-substring-search-begin() {
     _history_substring_search_query=$BUFFER
 
     #
-    # $BUFFER contains the text that is in the command-line currently.
-    # we put an extra "\\" before meta characters such as "\(" and "\)",
-    # so that they become "\\\(" and "\\\)".
+    # compose search pattern
     #
-    local escaped_query=${BUFFER//(#m)[\][()|\\*?#<>~^]/\\$MATCH}
+    if [[ -n $HISTORY_SUBSTRING_SEARCH_FUZZY ]]; then
+      #
+      # `=` split string in arguments
+      #
+      _history_substring_search_query_parts=(${=_history_substring_search_query})
+    else
+      _history_substring_search_query_parts=(${_history_substring_search_query})
+    fi
 
     #
-    # Find all occurrences of the search query in the history file.
+    # Join query parts with wildcard character '*' as seperator
+    # `(q)` escape special characters within string
+    # `(j:CHAR:)` join array to string with CHAR as seperator
+    # This is done in two steps to prevent quoting the wildcard character '*'
+    #
+    local search_pattern="*${(j:*:)${(q)_history_substring_search_query_parts[@]}}*"
```

Comment:
You  can just compute a separate version of this in the branches of the if above.